File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/TestClassShouldHaveTestMethod.MsTest.cs
using System;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace Tests.Diagnostics
{
    [TestClass]
    class ClassTest2 // Noncompliant
//        ^^^^^^^^^^
    {
    }

    [TestClass]
    class ClassTest4 // Noncompliant
    {
        public void Foo() { }
    }

    class ClassTest5
    {
        public void Foo() { }
    }

    class ClassTest6
    {
        public void Foo() { }
    }

    [TestClass]
    class ClassTest9
    {
        [TestMethod]
        public void Foo() { }
    }

    [TestClass]
    class ClassTest12
    {
        [DataTestMethod]
        [DataRow(1)]
        public void Foo(int i) { }
    }

    [TestClass]
    public abstract class MyCommonCode1
    {
    }

    [TestClass]
    public class MySubCommonCode1 : MyCommonCode1 // Noncompliant
    {
    }

    public class MySubCommonCode11 : MyCommonCode1 // Compliant
    {
    }

    [TestClass]
    public abstract class TestFooBase
    {
        [TestMethod]
        public void Foo_WhenFoo_ExpectsFoo() { }
    }

    // See https://github.com/SonarSource/sonar-csharp/issues/1196
    [TestClass]
    public class TestSubFoo : TestFooBase // Compliant - base abstract and at least 1 test in base
    {

    }
}

namespace TestSetupAndCleanupAttributes
{
    using Microsoft.VisualStudio.TestTools.UnitTesting;

    // Regression tests for Bug 1486: https://github.com/SonarSource/sonar-csharp/issues/1486
    // Don't raise for classes that test setup or cleanup attributes

    [TestClass]
    public class SetupAttributes1
    {
        [AssemblyInitialize]
        public static void BeforeTests(TestContext context)
        {
        }
    }

    [TestClass]
    public static class SetupAttributes2
    {
        [AssemblyCleanup]
        public static void AfterTests()
        {
        }
    }


    [TestClass]
    public static class SetupAttributes3
    {
        [AssemblyInitialize]
        public static void BeforeTests(TestContext context)
        {
        }

        [AssemblyCleanup]
        public static void AfterTests()
        {
        }
    }
}

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/TestClassShouldHaveTestMethod.MsTest.vb
Imports System
Imports Microsoft.VisualStudio.TestTools.UnitTesting

Namespace Tests.Diagnostics
    <TestClass>
    Class ClassTest2
    End Class

    <TestClass>
    Class ClassTest4
        Public Sub Foo()
        End Sub
    End Class

    Class ClassTest5
        Public Sub Foo()
        End Sub
    End Class

    Class ClassTest6
        Public Sub Foo()
        End Sub
    End Class

    <TestClass>
    Class ClassTest9
        <TestMethod>
        Public Sub Foo()
        End Sub
    End Class

    <TestClass>
    Class ClassTest12
        <DataTestMethod>
        <DataRow(1)>
        Public Sub Foo(ByVal i As Integer)
        End Sub
    End Class

    <TestClass>
    Public MustInherit Class MyCommonCode1
    End Class

    <TestClass>
    Public Class MySubCommonCode1
        Inherits MyCommonCode1
    End Class

    Public Class MySubCommonCode11
        Inherits MyCommonCode1
    End Class

    <TestClass>
    Public MustInherit Class TestFooBase
        <TestMethod>
        Public Sub Foo_WhenFoo_ExpectsFoo()
        End Sub
    End Class

    <TestClass>
    Public Class Foo_WhenFoo_ExpectsFoo
        Inherits TestFooBase
    End Class
End Namespace

Namespace TestSetupAndCleanupAttributes
    <TestClass>
    Public Class SetupAttributes1
        <AssemblyInitialize>
        Public Shared Sub BeforeTests(ByVal context As TestContext)
        End Sub
    End Class

    <TestClass>
    Module SetupAttributes2
        <AssemblyCleanup>
        Sub AfterTests()
        End Sub
    End Module

    <TestClass>
    Module SetupAttributes3
        <AssemblyInitialize>
        Sub BeforeTests(ByVal context As TestContext)
        End Sub

        <AssemblyCleanup>
        Sub AfterTests()
        End Sub
    End Module
End Namespace
